### PR TITLE
[XLA:GPU] BatchNorm lowering overhaul for cudnn fast fp16 kernels and act+side input fusion

### PR DESCRIPTION
**Pull request by AyanmoI**
**Problem Definition:**
1. Currently, BatchNorm, when calling into cuDNN, forces the data layout to be NCHW (not by transforms but by squashing physical dimensions) irrespective of their original layout. This leads to calls into suboptimal BatchNorm kernels in cuDNN resulting in drastic performance regression for NHWC models. Note that latest cuDNN versions have capability of calling into fast semi-persistent NHWC BatchNorm fp16 kernels. 
2.  In addition, cuDNN allows fusion of BN+Act and BN+Add+Act (implemented in this PR) that improves BN Training performance (over BatchNorm expansion).  BN Backward fusion has not been implemented yet.
Note, whether or not XLA calls into fused cuDNN BN depends on whether Grappler remaps FusedBatchNormV3 to FusedBatchNormEx. CuDNN FusedBatchNormEx API (which allows fusion)  is only valid
a) NHWC datatype,
b) DT_HALF,
c) only ReLU activation,
d) Channel dim has to be a multiple of 4, and
e) BatchNormSpatial Persistent mode.

**Implementation:**
Calling into the fast semi-persistent  and fused kernels require an overhaul of the BatchNorm lowering since these require additional reserve space (transfer bits from the forward to the backward BatchNorm op) and work space (scratch space).
This PR modifies BatchNorm lowering from TF2XLA bridge to rewriter optimization to thunks in order to make sure that the performant kernels get called.

### CODE REVIEW

File: tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td [hunk 1]
```diff
@@ -1227,8 +1228,12 @@ def HLO_BatchNormTrainingOp : HLO_Op<"batch_norm_training", [NoSideEffect]> {
     HLO_Tensor:$operand,
     HLO_Tensor:$scale,
     HLO_Tensor:$offset,
+    HLO_Tensor:$side_input,
     F32Attr:$epsilon,
-    I64Attr:$feature_index
+    I64Attr:$feature_index,
+    I64Attr:$reserve_space_size,
+    BoolAttr:$use_reserve_space,
```
Comment (sanjoy) [inline tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td:1235]: I could be missing something here, but we don't allow backend specific concepts at the HLO layer.  Can you do this optimization in a way that does not require adding these inputs to BatchNorm?

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td (hunk 1 above)
Comment (AyanmoI) [reply tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td:1235]: The additional attributes introduced here are optional and the HLO will should work with all backends. In this implementation, I am leveraging the TensorFlow BN op's additional reserve space output and the data flow is based on the correct lowering of graph-def at the bridge. Incase there is no side_input and/or activation, the HLO for BatchNorm remains the same.

**Comment by gbaned** [comment]
@AyanmoI  This PR is in draft, any update on this? Please. Thanks!

**Comment by gbaned** [comment]
@AyanmoI  Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@AyanmoI Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@AyanmoI Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@AyanmoI  Any update on this PR? Please. Thanks!
